sql: TestCancelDistSQLQuery failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/ee5696d9afdc7cda7f7cea6520928f98af58e335

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestCancelDistSQLQuery PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=918504&tab=buildLog

```
=== RUN   TestCancelDistSQLQuery
W180921 05:43:18.421989 92063 server/status/runtime.go:310  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180921 05:43:18.452625 92063 server/server.go:852  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180921 05:43:18.452763 92063 base/addr_validation.go:279  [n?] server certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I180921 05:43:18.452783 92063 base/addr_validation.go:319  [n?] web UI certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I180921 05:43:18.455134 92063 server/config.go:493  [n?] 1 storage engine initialized
I180921 05:43:18.455161 92063 server/config.go:496  [n?] RocksDB cache size: 128 MiB
I180921 05:43:18.455170 92063 server/config.go:496  [n?] store 0: in-memory, size 0 B
I180921 05:43:18.461537 92063 util/stop/stopper.go:537  [n?] quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
I180921 05:43:18.462217 92063 server/node.go:374  [n?] **** cluster 5d5402a8-d0f8-4a79-a631-b2000ad66a6e has been created
I180921 05:43:18.462248 92063 server/server.go:1414  [n?] **** add additional nodes by specifying --join=127.0.0.1:39833
I180921 05:43:18.462427 92063 gossip/gossip.go:408  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:39833" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:12 > build_tag:"v2.1.0-beta.20180917-206-gee5696d" started_at:1537508598462352889 
I180921 05:43:18.463339 92135 gossip/gossip.go:940  [n1] gossip connectivity
I180921 05:43:18.463911 92063 storage/store.go:1559  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180921 05:43:18.464038 92063 server/node.go:477  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=7.0 KiB), ranges=1, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=7183.00 p25=7183.00 p50=7183.00 p75=7183.00 p90=7183.00 pMax=7183.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I180921 05:43:18.464106 92063 storage/stores.go:242  [n1] read 0 node addresses from persistent storage
I180921 05:43:18.464195 92063 server/node.go:700  [n1] connecting to gossip network to verify cluster ID...
I180921 05:43:18.469398 92063 server/node.go:725  [n1] node connected via gossip and verified as part of cluster "5d5402a8-d0f8-4a79-a631-b2000ad66a6e"
I180921 05:43:18.469431 92063 server/node.go:549  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I180921 05:43:18.469637 92063 server/status/recorder.go:611  [n1] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180921 05:43:18.469658 92063 server/server.go:1805  [n1] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180921 05:43:18.469719 92063 server/server.go:1551  [n1] starting https server at 127.0.0.1:38199 (use: 127.0.0.1:38199)
I180921 05:43:18.469732 92063 server/server.go:1553  [n1] starting grpc/postgres server at 127.0.0.1:39833
I180921 05:43:18.469743 92063 server/server.go:1554  [n1] advertising CockroachDB node at 127.0.0.1:39833
I180921 05:43:18.478691 92356 sql/event_log.go:126  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I180921 05:43:18.490833 92143 sql/event_log.go:126  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:2.0-12 User:root}
I180921 05:43:18.498498 92370 sql/event_log.go:126  [n1,intExec=disableNetTrace] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:root}
I180921 05:43:18.518233 92390 sql/event_log.go:126  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:4ce1828c-8527-4590-a3b5-c755fb570836 User:root}
I180921 05:43:18.523896 92360 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I180921 05:43:18.526771 92363 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I180921 05:43:18.528223 92063 server/server.go:1607  [n1] done ensuring all necessary migrations have run
I180921 05:43:18.528248 92063 server/server.go:1610  [n1] serving sql connections
I180921 05:43:18.531767 92367 server/server_update.go:67  [n1] no need to upgrade, cluster already at the newest version
I180921 05:43:18.532150 92369 sql/event_log.go:126  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:39833} Attrs: Locality: ServerVersion:2.0-12 BuildTag:v2.1.0-beta.20180917-206-gee5696d StartedAt:1537508598462352889 LocalityAddress:[]} ClusterID:5d5402a8-d0f8-4a79-a631-b2000ad66a6e StartedAt:1537508598462352889 LastUp:1537508598462352889}
W180921 05:43:18.534904 92063 server/status/runtime.go:310  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180921 05:43:18.540647 92063 server/server.go:852  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180921 05:43:18.540845 92063 base/addr_validation.go:279  [n?] server certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I180921 05:43:18.540865 92063 base/addr_validation.go:319  [n?] web UI certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I180921 05:43:18.548538 92063 server/config.go:493  [n?] 1 storage engine initialized
I180921 05:43:18.548562 92063 server/config.go:496  [n?] RocksDB cache size: 128 MiB
I180921 05:43:18.548571 92063 server/config.go:496  [n?] store 0: in-memory, size 0 B
W180921 05:43:18.548620 92063 gossip/gossip.go:1516  [n?] no incoming or outgoing connections
I180921 05:43:18.548687 92063 server/server.go:1416  [n?] no stores bootstrapped and --join flag specified, awaiting init command.
I180921 05:43:18.563757 92307 gossip/client.go:129  [n?] started gossip client to 127.0.0.1:39833
I180921 05:43:18.564209 92469 gossip/server.go:226  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:45063}
I180921 05:43:18.577202 92443 gossip/gossip.go:940  [n?] gossip connectivity
  n1 [sentinel];
I180921 05:43:18.577443 92063 server/node.go:700  [n?] connecting to gossip network to verify cluster ID...
I180921 05:43:18.577501 92063 server/node.go:725  [n?] node connected via gossip and verified as part of cluster "5d5402a8-d0f8-4a79-a631-b2000ad66a6e"
I180921 05:43:18.582348 92063 server/node.go:429  [n?] new node allocated ID 2
I180921 05:43:18.582502 92063 gossip/gossip.go:408  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:45063" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:12 > build_tag:"v2.1.0-beta.20180917-206-gee5696d" started_at:1537508598582396171 
I180921 05:43:18.582634 92063 storage/stores.go:242  [n2] read 0 node addresses from persistent storage
I180921 05:43:18.582708 92063 storage/stores.go:261  [n2] wrote 1 node addresses to persistent storage
I180921 05:43:18.583022 92443 gossip/gossip.go:940  [n2] gossip connectivity
  n1 [sentinel];
  n2 -> n1;
I180921 05:43:18.584043 92098 storage/stores.go:261  [n1] wrote 1 node addresses to persistent storage
I180921 05:43:18.584168 92135 gossip/gossip.go:940  [n1] gossip connectivity
  n1 [sentinel];
  n2 -> n1;
I180921 05:43:18.585791 92063 server/node.go:675  [n2] bootstrapped store [n2,s2]
I180921 05:43:18.585968 92063 server/node.go:549  [n2] node=2: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I180921 05:43:18.586295 92063 server/status/recorder.go:611  [n2] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180921 05:43:18.586332 92063 server/server.go:1805  [n2] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180921 05:43:18.586467 92063 server/server.go:1551  [n2] starting https server at 127.0.0.1:46151 (use: 127.0.0.1:46151)
I180921 05:43:18.603195 92063 server/server.go:1553  [n2] starting grpc/postgres server at 127.0.0.1:45063
I180921 05:43:18.603216 92063 server/server.go:1554  [n2] advertising CockroachDB node at 127.0.0.1:45063
I180921 05:43:18.605456 92063 server/server.go:1607  [n2] done ensuring all necessary migrations have run
I180921 05:43:18.605490 92063 server/server.go:1610  [n2] serving sql connections
I180921 05:43:18.631673 92453 sql/event_log.go:126  [n1,client=127.0.0.1:38032,user=root] Event: "create_database", target: 52, info: {DatabaseName:test Statement:CREATE DATABASE IF NOT EXISTS test User:root}
I180921 05:43:18.636471 92453 sql/event_log.go:126  [n1,client=127.0.0.1:38032,user=root] Event: "create_table", target: 53, info: {TableName:test.public.nums Statement:CREATE TABLE test.public.nums (num INT) User:root}
I180921 05:43:18.640868 92453 storage/replica_command.go:298  [n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /Table/53/1/50 [r2]
I180921 05:43:18.646454 92599 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      node.Node: writing summary
1      [async] closedts-rangefeed-subscriber
1      [async] auto-upgrade
W180921 05:43:18.647434 92476 internal/client/txn.go:532  [n2,intExec=read-setting] failure aborting transaction: node unavailable; try another peer; abort caused by: context canceled
W180921 05:43:18.647494 92563 server/node.go:848  [n2,summaries] error recording initial status summaries: failed to send RPC: sending to all 1 replicas failed; last error: <nil> rpc error: code = Unavailable desc = transport is closing
I180921 05:43:18.647532 92599 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      [async] closedts-rangefeed-subscriber
1      [async] auto-upgrade
I180921 05:43:18.646575 92598 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
W180921 05:43:18.647154 92568 internal/client/txn.go:532  [n2,hb] failure aborting transaction: node unavailable; try another peer; abort caused by: result is ambiguous (error=rpc error: code = Unavailable desc = transport is closing [exhausted])
I180921 05:43:18.647571 92599 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      [async] auto-upgrade
I180921 05:43:18.647593 92568 storage/node_liveness.go:817  [n2,hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (error=rpc error: code = Unavailable desc = transport is closing [exhausted])
W180921 05:43:18.647603 92475 internal/client/txn.go:532  [n2,intExec=show-version] failure aborting transaction: node unavailable; try another peer; abort caused by: read-setting: context canceled
I180921 05:43:18.647650 92471 server/server_update.go:64  [n2] failed attempt to upgrade cluster version, error: show-version: context canceled
I180921 05:43:18.647668 92599 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
W180921 05:43:18.647674 92568 internal/client/txn.go:532  [n2,hb] failure aborting transaction: node unavailable; try another peer; abort caused by: node unavailable; try another peer
W180921 05:43:18.647693 92568 storage/node_liveness.go:494  [n2,hb] failed node liveness heartbeat: node unavailable; try another peer
W180921 05:43:18.647251 92562 ts/db.go:195  [n2,ts-poll] error writing time series data: failed to send RPC: sending to all 1 replicas failed; last error: <nil> rpc error: code = Unavailable desc = transport is closing
W180921 05:43:18.647360 92473 server/node.go:945  [n2] node=2: unable to log node_join event: log-event: failed to send RPC: sending to all 1 replicas failed; last error: <nil> rpc error: code = Unavailable desc = transport is closing
--- FAIL: TestCancelDistSQLQuery (0.24s)
	run_control_test.go:245: pq: cannot up-repliate to s2; missing gossiped StoreDescriptor

```

--- a/pkg/sql/run_control_test.go
+++ b/pkg/sql/run_control_test.go
@@ -236,14 +236,16 @@ func TestCancelDistSQLQuery(t *testing.T) {
 	if _, err := conn1.Exec("ALTER TABLE nums SPLIT AT VALUES (50)"); err != nil {
 		t.Fatal(err)
 	}
-	// Make the second node the leaseholder for the first range to distribute
-	// the query.
-	if _, err := conn1.Exec(fmt.Sprintf(
-		"ALTER TABLE nums EXPERIMENTAL_RELOCATE VALUES (ARRAY[%d], 1)",
-		tc.Server(1).GetFirstStoreID(),
-	)); err != nil {
-		t.Fatal(err)
-	}
+
+	// Make the second node the leaseholder for the first range to distribute the
+	// query. This may have to retry if the second store's descriptor has not yet
+	// propagated to the first store's StorePool.
+	testutils.SucceedsSoon(t, func() error {
+		_, err := conn1.Exec(fmt.Sprintf(
+			"ALTER TABLE nums EXPERIMENTAL_RELOCATE VALUES (ARRAY[%d], 1)",
+			tc.Server(1).GetFirstStoreID()))
+		return err
+	})
 
 	// Run queryToCancel to be able to get an estimate of how long it should
 	// take. The goroutine in charge of cancellation will sleep a random

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -1131,7 +1131,7 @@ func (s *Store) AdminRelocateRange(
 			for _, candidate := range candidates {
 				store, ok := storeMap[candidate.StoreID]
 				if !ok {
-					return fmt.Errorf("cannot up-repliate to s%d; missing gossiped StoreDescriptor", candidate.StoreID)
+					return fmt.Errorf("cannot up-replicate to s%d; missing gossiped StoreDescriptor", candidate.StoreID)
 				}
 				candidateDescs = append(candidateDescs, *store)
 			}

